org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/scoping/batch/NestedTypesScope.java

```diff
@@ -71,11 +74,17 @@ protected void addDescriptions(JvmDeclaredType type, JvmType declarator, List<IE
 		Iterator<JvmDeclaredType> iterator = nestedTypeDeclarators.iterator();
 		String simpleName = name.getFirstSegment();
 		List<IEObjectDescription> result = null;
-		while(iterator.hasNext()) {
-			JvmDeclaredType declarator = iterator.next();
-			Iterable<JvmDeclaredType> nestedTypes = declarator.findAllNestedTypesByName(simpleName);
-			if (name.getSegmentCount() == 1 && nestedTypes.iterator().hasNext()) {
-				result = addDescriptions(name, 0, nestedTypes, result);
+		if (name.getSegmentCount() == 1) {
+			Set<JvmDeclaredType> nestedTypesSet = new LinkedHashSet<>();
+			while(iterator.hasNext()) {
+				JvmDeclaredType declarator = iterator.next();
+				Iterable<JvmDeclaredType> nestedTypes = declarator.findAllNestedTypesByName(simpleName);
+				if (nestedTypes.iterator().hasNext()) {
```

Comment:
Can be simplified, since iterator.hasNext() will be checked in addAll below anyways.